lib/livebook/utils.ex

```diff
@@ -27,6 +27,50 @@ defmodule Livebook.Utils do
     :crypto.strong_rand_bytes(42) |> Base.url_encode64() |> String.to_atom()
   end
 
+  @doc """
+  Generates a random binary id that includes node information.
+
+  ## Format
+
+  The id is formed from the following binary parts:
+
+    * 16B - hashed node name
+    * 9B - random bytes
+
+  The binary is base32 encoded.
+  """
+  @spec random_node_aware_id() :: id()
+  def random_node_aware_id() do
+    node_part = node_hash(node())
+    random_part = :crypto.strong_rand_bytes(9)
+    binary = <<node_part::binary, random_part::binary>>
+    # 16B + 9B = 25B is suitable for base32 encoding without padding
+    Base.encode32(binary, case: :lower)
+  end
+
+  # Note: the result is always 16 bytes long
+  defp node_hash(node) do
+    content = Atom.to_string(node)
+    :crypto.hash(:md5, content)
+  end
+
+  @doc """
+  Extracts node name from the given node aware id.
+
+  The node in question must be connected, otherwise it won't be found.
+  """
+  @spec node_from_node_aware_id(id()) :: {:ok, node()} | :error
+  def node_from_node_aware_id(id) do
+    binary = Base.decode32!(id, case: :lower)
+    <<node_part::binary-size(16), _random_part::binary-size(9)>> = binary
+
+    known_nodes = [node() | Node.list()]
+
+    Enum.find_value(known_nodes, :error, fn node ->
+      node_hash(node) == node_part && {:ok, node}
+    end)
+  end
```

Comment:
@josevalim encoding node with base32 would result in variable-length and overly long ids, so instead we store md5 hash and then "reverse engineer" what node it is :cat: